### ISSUE TITLE
JCR-WebDAV migration: Need to modify PropertyFactory

### ISSUE THREAD

**Issue by TheAntimist** [caldav4j/caldav4j#59]
@ioggstream There's two things here, one PropertyFactory in `org.apache.webdav.lib.properties.PropertyFactory` another in `org.osaf.caldav4j.model.util.PropertyFactory`. I feel that both of these can be replaced by `org.apache.jackrabbit.webdav.property.DavPropertyName` and the other by `org.apache.jackrabbit.webdav.property.DavPropertyNameSet`. What do you suggest?

I mean for replacing the caldav4j.PropertyFactory, seems unecessary, but to we need to replace the jakarta slide one.

From what I understand, the register function, seems to be a sort of a cache, for properties most frequently used. To replace it would be fairly simple, if that is all that it does?

**Comment by ioggstream** [comment]
@TheAntimist good, you opened a new tkt.

Ok, replace with jackrabbit as much as you can: the less code we have in caldav4j the better.

I don't remember using directly the register function...

**Comment by TheAntimist** [comment]
@ioggstream Sigh, two places, copy pasting here again. And sticking to this one. 

DavPropertyName, has this function:

    public Element toXml(Document document)

So, I am assuming yes.

**Comment by ioggstream** [comment]
Implement PropertyFactory test cases and check if they works ;)

**Comment by TheAntimist** [comment]
@ioggstream Let me think over this one, because if I change one, I change all. Like for example a lot of the PropProperty, can be converted to DavProperty in jcr. Along with that I also, would have to look into changing the Response type, from all those OutputsDOM to, from what I think most of the Responses in JCR is MultiStatus. It's sort of a big issue, and I feel that I don't have the broader view. Let me analyze a part of the code, then it'd be easier.

**Comment by ioggstream** [comment]
The CaldavXMLResponseMethodBase is one of the big issue indeed. If you find a way to reuse existing code to:

  - create queries
  - parse the XMLResponse 

You're done.

**Comment by TheAntimist** [comment]
@ioggstream I'll have to look into it. Thanks for the mentioning those, might help.

**Comment by TheAntimist** [comment]
@ioggstream Having looked at what jackrabbit does for it's implementation as a base, we can perhaps do something similar, which would yield code, similar to jackrabbit, and be leaner and cleaner. From what I understand:

 - Requests: If a class needs to properties or any specific property to be set, they use an extra class, which for example in ``MkCalendarMethod``, instead of internalizing, they say that the user should set all their properties in a class similar to ``MkCalendarInfo``, which implements, XMLSerializable, i.e. it has a toXml(), function. which in the constructor, can then be used to setRequestBody(XML).

- Parsing Responses: This is can be handled in DavMethodBase and HTTPMethodBase, which has these functions
   * processResponseHeaders
   * processResponseBody
   * processResponseHeaders
which can then be overridden to work.

Also, responses, can by default be outputted using MultiStatus. And some Properties and so on, will have to migrated to the jcr properties, if feasible, otherwise, this method, would be simpler and better, also, cleaner. By seperating out the code for Methods and the information, they pass, like ReportMethod, has a ReportInfo, class in jcr, which sets all the data for the report and then sends that to the Method.

What do you think? Should move ahead with this scenario, which keeps compatibility with jcr and makes the code cleaner, but breaks the current implementation, hopefully, not by much.

**Comment by ioggstream** [comment]
I suggest to:

  - identify two tests to be implemented with those parsing classes: one easy and one hard
  - make the test work
  - identify the effort.

I suggest too to merge all your patches to the jackrabbit branch before starting all that work.

Does it make sense to you?

**Comment by TheAntimist** [comment]
@ioggstream Just to be clear on this, what you are saying is that I convert, like one or two classes using the method, which I mentioned. Merge all that to my jackrabbit branch and check for the tests, which you mentioned right?

I'm thinking of going with probably MkCalendar and Put and/or Get. To test these cases. That seem like a decent plan?

**Comment by TheAntimist** [comment]
@ioggstream In PropFindMethod, could you tell me what the ``errorMap = new HashMap<QName, Error>();`` is for?

I seem to not understand why do we parse the new responses and save all the errors in this map?

**Comment by TheAntimist** [comment]
@ioggstream Since, I know from the comments of this article, that you are aware of this: http://blogs.nologin.es/rickyepoderi/index.php?/archives/15-Introducing-CalDAV-Part-II.html

My plan was to originally, Info classes for the methods, similar to the article. Like for example use the ReportInfo class for ReportMethod, and add the other necessary or overridden methods, for like calendar-query and the normal ReportMethod. Implemented something like this for Propfind, but using the DavPropertyNameSet, as the class which sets all the properties to find, for each, calendar or event. And the tests, were easy to implement, because it's an extension of the jcr-webdav, code.

Problem is as I said, breaks, all previous implementations, because we the jcr-webdav classes. So for example, unlike the old PropFindMethod, the Properties, are not enumeration, instead, they are returned as DavPropertySet. Also, the whole requestBody is either returned as the MultiStatus class or as Document.

Should I go ahead with this?

**Comment by ioggstream** [comment]
@TheAntimist

 -  the errorMap is because we want to parse everything before replying.

> Properties, are not enumeration, instead, they are returned as DavPropertySet

Again, if I remember correctly that's the point where the work was stopped: maybe creating  an utility class to convert enumerations in DavPropertySet  could help.

Did you already do some modifications?

**Comment by TheAntimist** [comment]
@ioggstream Alright, thanks for the errormap, I might seperate it out onto it's own class.

What about using Prop, the property set element in Caldav4j and PropProperty, the which would be properties for Caldav4j. and returning it as either one of these. Because these can be made to extend the DavPropertySet and DavProperty, while still keeping their API, along with additional functions. Just that conversion to Xml, will now have to be handled in it's own class with a toXml, function. Other than that, it should be fine.

I've been working on PropFind, and since the contructor, takes in the Properties for which to apply the Propfind for, that's one change, but I feel this can be handled through, Prop and PropProperty as I stated earlier.

**Comment by TheAntimist** [comment]
Ok, here's an example of what PropFind would look like after migration: https://github.com/TheAntimist/caldav4j/blob/c8b95889092ac0a84820ef1e4ef829d8e355c020/src/main/java/org/osaf/caldav4j/methods/NewPropFindMethod.java

It's mostly completed, tried keeping it as close as possible to the original, while using the Jackrabbit classes. Though, lots of changes, there.

**Comment by TheAntimist** [comment]
@ioggstream Here's the modified NewPropFindTest. https://github.com/TheAntimist/caldav4j/blob/c9d3c8850c65f1878d5cbc7645a6ccec5e5b64d0/src/test/java/org/osaf/caldav4j/methods/NewPropFindTest.java

**Comment by TheAntimist** [comment]
@ioggstream I need to modify the OutputsDomBase, and noticed, that most of these classes which extend this use a variable something like this: ``caldavNamespaceQualifier``. But this variable seems redundant, since, most of these classes have a manually set ``getNamespaceURI()``. Would it be better if I remove it? or make another constructor, which gives the automatic value to the current one, without, removing the old constructor.

**Comment by TheAntimist** [comment]
@ioggstream Mostly done with the migration of the Report Method, I'll have to still run up couple of tests, for making sure there's no trouble. Anyway, here's a list of changes to expect:

 - For settings properties in the new method, we use DavPropertyNameSet
 - OutputsDom, now extends XmlSerializable. Thus, removing/renaming the createNewDocument or a toXml, function. which is an extension of the previous functions.
 - Prop and PropProperty, now extend DavProperty, but with children and attributes.
 - PropProperty can be assigned to DavProperty, but not the other way round. Basic inheritance.
 - Using the class Namespace instead, of the qualifier and the URI, seperately. Added the defaults to Constants.
 - AclProperty from jakarta, now migrated over to AclProperty in jcr, and Ace[] is now a ``List<AclProperty.Ace>``
 - Responses, as stated earlier are MultiStatus, or MultiStatusResponses.
 - With respect to ReportMethod, I have merged both the Calendar and CalDAV ReportMethods, into one class, which provides both their functionality.
 - SimpleDOMOutputtingObject, is no longer needed due to PropProperty. But kept it for compatibility's sake.
 - ReportMethod, now takes in jackrabbit's ReportInfo classes, while maintaining compatibility with jcr component. So, CalendarQuery, CalendarMultiget, FreeBusyQuery, now extend ReportInfo.
 - One major change is that you can't modify the properties, added to the method, as in the jcr implementation, they use, ``private final``.
 - CalDAVXMLResponseMethod, no longer needed. Since, we directly extend the ReportMethod.
 - QName, replaced by DavPropertyName.
 - Privilege, to be modified to the jcr Privilege class.

The only change which might keep compatibility with the old ways, is that CalDAVResponse, can be made to extend MultiStatusResponse. Might help people already, using. Running tests on these for now, should have them uploaded soon. As much as I tried to not break compatibility and API's, there were still a need to make some of these changes, like Property's and so on.

**Comment by TheAntimist** [comment]
@ioggstream I have just now commited lots of files, with all those changes previously mentioned here ae6a9d2553b47a25abd21b7ecbbce168d31da246 .What do you think? 

if you have any queries or anything let me know.

And simply put, PropProperty, now extends DavProperty. Prop extends PropProperty. OutputsDom extends XmlSerializable, i.e. has a toXml() function. CalDAVReportMethod, centralizes the reportmethod for both Calendar and CalDAVReport Method.

**Comment by TheAntimist** [comment]
@ioggstream Moving forward, I've been working on the Response Handling aspects, and here are the conclusions, I've been able to draw, by Trial and Error.

 - CalDAVResponse, even if it extends MultiStatusResponse, you still cannot do this: ``CalDAVResponse calDAVResponse = (CalDAVResponse) multistatusResponse;``. Cast Exception is thrown due to basic inheritance problem.

Though one possible solution is to have static methods for conversions of Properties and so on in the class. Otherwise, there won't be any other use for this class.
 - CalendarDataProperty, can extend PropProperty or DavProperty, but still won't be able to do this: `` return (CalendarDataProperty) getProperty(PropertyName);``. It again throws a Cast exception, similar to the previous Point.
Only way to keep it is again to have to static method for converting normal Property to Calendar.
 - CalDAVResource can be changed from CalDavResponse to MultiStatusResponse.
 - TicketResponse,  no change.
 - Principal, use/extend the Principal in jcr
 - As stated earlier, no more enumeration of Responses is possible, because the Response Body is not MultiStatus.

Any suggestions on how to handle the CalDAVResponse would be appreciated.

**Comment by ioggstream** [comment]
Today I will look the code.

Maybe we could reuse existing code to reimplement from scratch the new
classes...

**Comment by TheAntimist** [comment]
@ioggstream Looking forward to what you think about the situation.

**Comment by ioggstream** [comment]
@TheAntimist trying building your branch. How will you replace the QName in CaldavConstants ?

**Comment by TheAntimist** [comment]
@ioggstream Either use PropProperty or DavPropertyName, to keep compatibility I used PropProperty, but either can be used. Here's the ETAG ``PROP_GETETAG``, instead of ``QName_ETAG``

**Comment by ioggstream** [comment]
Ok, then we can get rid of QNAME_xxx ...

**Comment by ioggstream** [comment]
@TheAntimist it seems the basic stuff is right. Please create a new branch where you get rid of old classes (eg. slide) and use new methods instead of old ones. (Eg. delete CaldavReportMethod, rename  NewCaldavReportMethod -> CaldavReportMethod)

**Comment by ioggstream** [comment]
@TheAntimist in the jcr version we'll remove the deprecated CaldavCalendarCollection class and leave only the CaldavCollection class.

**Comment by TheAntimist** [comment]
@ioggstream I'll make a new branch today with all those that you mentioned. Other than that, should I temporarily use MultiStatus and MultiStatusResponse in the CaldavCollection collection class, or wait a bit.

**Comment by TheAntimist** [comment]
@ioggstream Just one thing, I've had a bit of question about. With PropProperty, Prop there along with DavPropertyName and DavProperty. Sometimes it might get confusing, which to use where. For CalDAVReportMethod, I've made it possible to use any of the four for the Property, but, with the MultiStatusResponse only returning DavProperty, if the users want to use the PropProperty, they'll find themselves, not being able to use it. It's why, for NewPropFindMethod, I haven't returned anything other than DavProperty, but I can modify it for adding PropProperty and Prop to the method properties.

Anyway, back to working on the branch.

**Comment by TheAntimist** [comment]
@ioggstream Refer commit 6b90ffa03af814474416d7aef4e2d28ffd6ad9f6 for the new update in my fork for branch jcr-dev.

I have removed all of the Response handling and made it to MultiStatus Responses. Haven't tested everything, except for Tickets, which now work as well. I believe that PropFind and CalDAVReport, should continue to work.

List of Changes:

 * Deleted all the Slide Methods
 * Removed CalDAVCalendarCollection class
 * TicketDiscoveryProperty now, has a constructor which takes in MultiStatusResponse or MultiStatus
 * Added more methods to Propfind and CalDAVReport
 * Migrated all responses to either MultiStatus or MultiStatusResponse.
 * CalendarDataProperty, now only has Static methods to get Etag and Calendar from Response.
 * Privileges now refactored to CalDAVPrivilege due to conflict with jcr Privilege and registering them.
 * CalDAVResource, updated to use MultiStatusResponses.


Comments are appreciated.

**Comment by TheAntimist** [comment]
@ioggstream Now removed all traces of jakarta-slide, except for in the case of Principal, which for now I have deprecated, in favor of jackrabbit-webdav Principal.

One update I made in XMLUtils is to replace in ``toPrettyXML`` instead of XMLSerializer to ``Transformer`` as now XMLSerializer is deprecated. Also, tried getting most of the XML related functions streamlined with jcr.

Removed all calDavNamespaceQualifier and so on, from the constructors and body.

Other than that I have mostly run all the tests including the CalDAVCollectionTests and the rest. Everything, seems to be working fine.

Please refer commit 5a4e211ec07d4bce027b840fc8ae9dfbbf63386b
Any comments and further discussion on CalDAVResponse and Response Handling is appreciated.

**Comment by TheAntimist** [comment]
@ioggstream I just found out, that using ``ThreadLocal<CalendarBuilder>`` in MethodFactory causes a memory in a Java WebApplication like a .war running under Tomcat. Read more about it here: http://javarevisited.blogspot.in/2013/01/threadlocal-memory-leak-in-java-web.html

Any way to address this issue? Do we have to address this issue?

**Comment by TheAntimist** [comment]
Closing this issue, since, PropertyFactory is now removed from the jackrabbit branch. Making a new issue for the Web Application Memory leak.
